From d290157fa940282570b1803026e9f731a6699c14 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.com>
Date: Thu, 22 Feb 2024 15:19:48 +0000
Subject: [PATCH 120/126] arm64: dts: broadcom: Add in DRM pipeline to 2712.

Also increases CMA heap to 256MB as otherwise you have no space for
any framebuffers.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
---
 arch/arm64/boot/dts/broadcom/bcm2712.dtsi | 141 +++++++++++++++++++++-
 1 file changed, 135 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/boot/dts/broadcom/bcm2712.dtsi b/arch/arm64/boot/dts/broadcom/bcm2712.dtsi
index 292d7608cf..a41520d79c 100644
--- a/arch/arm64/boot/dts/broadcom/bcm2712.dtsi
+++ b/arch/arm64/boot/dts/broadcom/bcm2712.dtsi
@@ -24,7 +24,7 @@ atf@0 {
 
 		cma: linux,cma {
 			compatible = "shared-dma-pool";
-			size = <0x4000000>; /* 64MB */
+			size = <0x10000000>; /* 256MB */
 			reusable;
 			linux,cma-default;
 
@@ -33,8 +33,10 @@ cma: linux,cma {
 			 * ZONE_DMA32, that's not good enough for the BCM2711
 			 * as some devices can only address the lower 1G of
 			 * memory (ZONE_DMA).
+			 * Limit cma to the lower 768MB to allow room for
+			 * HIGHMEM on 32-bit
 			 */
-			alloc-ranges = <0x0 0x00000000 0x40000000>;
+			alloc-ranges = <0x0 0x00000000 0x30000000>;
 		};
 	};
 
@@ -72,6 +74,18 @@ clk_108MHz: clk-108M {
 		clock-output-names = "108MHz-clock";
 	};
 
+	hvs: hvs@107c580000 {
+		compatible = "brcm,bcm2712-hvs";
+		reg = <0x10 0x7c580000 0x1a000>;
+		interrupt-parent = <&disp_intr>;
+		interrupts = <2>, <9>, <16>;
+		interrupt-names = "ch0-eof", "ch1-eof", "ch2-eof";
+		clocks = <&firmware_clocks 4>, <&firmware_clocks 16>;
+		clock-names = "core", "disp";
+		//iommus = <&iommu4>;
+		status = "okay";
+	};
+
 	soc: soc {
 		compatible = "simple-bus";
 		#address-cells = <1>;
@@ -99,6 +113,53 @@ mailbox: mailbox@7c013880 {
 			#mbox-cells = <0>;
 		};
 
+		pixelvalve0: pixelvalve@7c410000 {
+			compatible = "brcm,bcm2712-pixelvalve0";
+			reg = <0x7c410000 0x100>;
+			interrupts = <GIC_SPI 101 IRQ_TYPE_LEVEL_HIGH>;
+			status = "okay";
+		};
+
+		pixelvalve1: pixelvalve@7c411000 {
+			compatible = "brcm,bcm2712-pixelvalve1";
+			reg = <0x7c411000 0x100>;
+			interrupts = <GIC_SPI 110 IRQ_TYPE_LEVEL_HIGH>;
+			status = "okay";
+		};
+
+		mop: mop@7c500000 {
+			compatible = "brcm,bcm2712-mop";
+			reg = <0x7c500000 0x20>;
+			interrupt-parent = <&disp_intr>;
+			interrupts = <1>;
+			status = "okay";
+		};
+
+		moplet: moplet@7c501000 {
+			compatible = "brcm,bcm2712-moplet";
+			reg = <0x7c501000 0x20>;
+			interrupt-parent = <&disp_intr>;
+			interrupts = <0>;
+			status = "okay";
+		};
+
+		disp_intr: interrupt-controller@7c502000 {
+			compatible = "brcm,bcm2711-l2-intc", "brcm,l2-intc";
+			reg = <0x7c502000 0x30>;
+			interrupts = <GIC_SPI 97 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-controller;
+			#interrupt-cells = <1>;
+			status = "okay";
+		};
+
+		dvp: clock@7c700000 {
+			compatible = "brcm,brcm2711-dvp";
+			reg = <0x7c700000 0x10>;
+			clocks = <&clk_108MHz>;
+			#clock-cells = <1>;
+			#reset-cells = <1>;
+		};
+
 		/*
 		 * This node is the provider for the enable-method for
 		 * bringing up secondary cores.
@@ -399,7 +460,7 @@ ddc0: i2c@7d508200 {
 			clock-frequency = <200000>;
 			#address-cells = <1>;
 			#size-cells = <0>;
-			status = "disabled";
+			status = "okay";
 		};
 
 		ddc1: i2c@7d508280 {
@@ -410,7 +471,7 @@ ddc1: i2c@7d508280 {
 			clock-frequency = <200000>;
 			#address-cells = <1>;
 			#size-cells = <0>;
-			status = "disabled";
+			status = "okay";
 		};
 
 		bscd: i2c@7d508300 {
@@ -492,7 +553,7 @@ aon_intr: interrupt-controller@7d510600 {
 			interrupts = <GIC_SPI 239 IRQ_TYPE_LEVEL_HIGH>;
 			interrupt-controller;
 			#interrupt-cells = <1>;
-			status = "disabled";
+			status = "okay";
 		};
 
 		pinctrl_aon: pinctrl@7d510700 {
@@ -614,6 +675,74 @@ bsc_pmu: i2c@7d544000 {
 			status = "disabled";
 		};
 
+		hdmi0: hdmi@7ef00700 {
+			compatible = "brcm,bcm2712-hdmi0";
+			reg = <0x7c701400 0x300>,
+			      <0x7c701000 0x200>,
+			      <0x7c701d00 0x300>,
+			      <0x7c702000 0x80>,
+			      <0x7c703800 0x200>,
+			      <0x7c704000 0x800>,
+			      <0x7c700100 0x80>,
+			      <0x7d510800 0x100>,
+			      <0x7c720000 0x100>;
+			reg-names = "hdmi",
+				    "dvp",
+				    "phy",
+				    "rm",
+				    "packet",
+				    "metadata",
+				    "csc",
+				    "cec",
+				    "hd";
+			resets = <&dvp 1>;
+			interrupt-parent = <&aon_intr>;
+			interrupts = <1>, <2>, <3>,
+				     <7>, <8>;
+			interrupt-names = "cec-tx", "cec-rx", "cec-low",
+					  "hpd-connected", "hpd-removed";
+			ddc = <&ddc0>;
+			dmas = <&dma32 10>;
+			dma-names = "audio-rx";
+			clocks = <&firmware_clocks 13>, <&firmware_clocks 14>, <&dvp 0>, <&clk_27MHz>;
+			clock-names = "hdmi", "bvb", "audio", "cec";
+			status = "okay";
+		};
+
+		hdmi1: hdmi@7ef05700 {
+			compatible = "brcm,bcm2712-hdmi1";
+			reg = <0x7c706400 0x300>,
+			      <0x7c706000 0x200>,
+			      <0x7c706d00 0x300>,
+			      <0x7c707000 0x80>,
+			      <0x7c708800 0x200>,
+			      <0x7c709000 0x800>,
+			      <0x7c700180 0x80>,
+			      <0x7d511000 0x100>,
+			      <0x7c720000 0x100>;
+			reg-names = "hdmi",
+				    "dvp",
+				    "phy",
+				    "rm",
+				    "packet",
+				    "metadata",
+				    "csc",
+				    "cec",
+				    "hd";
+			ddc = <&ddc1>;
+			resets = <&dvp 2>;
+			interrupt-parent = <&aon_intr>;
+			interrupts = <11>, <12>, <13>,
+				     <14>, <15>;
+			interrupt-names = "cec-tx", "cec-rx", "cec-low",
+					  "hpd-connected", "hpd-removed";
+			dmas = <&dma32 17>;
+			dma-names = "audio-rx";
+			clocks = <&firmware_clocks 13>, <&firmware_clocks 14>, <&dvp 1>, <&clk_27MHz>;
+			clock-names = "hdmi", "bvb", "audio", "cec";
+			status = "okay";
+		};
+
 	};
 
 	arm-pmu {
@@ -999,7 +1128,7 @@ v3d: v3d@2000000 {
 			clocks-names = "v3d";
 			interrupts = <GIC_SPI 250 IRQ_TYPE_LEVEL_HIGH>,
 				     <GIC_SPI 249 IRQ_TYPE_LEVEL_HIGH>;
-			status = "disabled";
+			status = "okay";
 		};
 
 		gicv2: interrupt-controller@7fff9000 {
-- 
2.43.0

